src/tools/kdb/external.c

```diff
@@ -0,0 +1,195 @@
+{
+	Key * tmp = keyNew ("/", KEY_END);
+	keySetBaseName (tmp, key);
+	Key * resultKey = ksLookup (binaries, tmp, KDB_O_NONE);
+	keyDel (tmp);
+	if (resultKey == NULL)
+	{
+		return NULL;
+	}
+	return keyString (resultKey);
+}
+
+
+int tryLoadExternal (char * commandName, KeySet * binaries)
+{
+	char * execPathPtr = getenv ("KDB_EXEC_PATH");
+	bool found = false;
+	char path[PATH_MAX] = { 0 };
+	char * saveptr;
+	struct stat buf;
+
+	if (execPathPtr)
+	{
+		char * execPath = strdup (execPathPtr);
+		char * token = strtok_r (execPath, ":", &saveptr);
+		while (token != NULL && !found)
+		{
+			snprintf (path, sizeof (path), "%s/%s", token, commandName);
+			found = stat (path, &buf) != -1;
+			token = strtok_r (NULL, ":", &saveptr);
+		}
+		elektraFree (execPath);
+	}
+
+	if (!found)
+	{
+		snprintf (path, sizeof (path), "%s/%s", BUILTIN_EXEC_FOLDER, commandName);
+		found = stat (path, &buf) != -1;
+	}
+
+	if (found)
+	{
+		Key * tmp = keyNew ("/", KEY_END);
+		keySetBaseName (tmp, commandName);
+		keySetString (tmp, path);
+		ksAppendKey (binaries, tmp);
+		return 0;
+	}
+	return 1;
+}
+
+int loadExternalSpec (KeySet * spec, KeySet * binaries, Key * errorKey)
+{
+	KDB * handle = kdbOpen (NULL, errorKey);
+	Key * baseKey = keyNew ("spec:" CLI_BASE_KEY, KEY_END);
+	KeySet * config = ksNew (0, KS_END);
+	if (kdbGet (handle, config, errorKey) == -1)
+	{
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERRORF (errorKey, "could not load '%s': %s", CLI_BASE_KEY, GET_ERR (baseKey));
+		keyDel (baseKey);
+		ksDel (config);
+		kdbClose (handle, errorKey);
+		return 1;
+	}
+	Key * cur = NULL;
+	KeySet * part = ksCut (config, baseKey);
+
+	for (elektraCursor it = 0; it < ksGetSize (part); ++it)
+	{
+		cur = ksAtCursor (part, it);
+		const Key * externalMeta = keyGetMeta (cur, "external");
+		const Key * externalBinary = keyGetMeta (cur, "bin");
+		bool isExternal = false;
+		if (externalBinary != NULL && externalMeta != NULL && elektraKeyToBoolean (externalMeta, &isExternal) && isExternal)
+		{ // add external spec and save path to binary
+			KeySet * externalCommandSpec = ksCut (part, cur);
+
+			Key * tmp = keyNew ("/", KEY_END);
+			keySetBaseName (tmp, keyBaseName (cur));
+			keySetString (tmp, keyString (externalBinary));
+			ksAppendKey (binaries, tmp);
+
+			ksAppend (spec, externalCommandSpec);
+			ksDel (externalCommandSpec);
+			it--;
+		}
+	}
+	ksDel (part);
+	ksDel (config);
+	kdbClose (handle, errorKey);
+	keyDel (baseKey);
+	return 0;
+}
+
+int runExternal (const char * bin, char ** argv, Key * errorKey)
+{
+	// the external program should think it was called directly
+	argv[1] = (char *) bin;
+
+	int status = 0;
+
+#ifdef _WIN32
+	STARTUPINFO si;
+	PROCESS_INFORMATION pi;
+
+	ZeroMemory (&si, sizeof (si));
+	si.cb = sizeof (si);
+	ZeroMemory (&pi, sizeof (pi));
+
+	// Construct command line string
+	char cmdline[MAX_PATH] = "";
+	for (int i = 1; argv[i]; ++i)
+	{
+		strcat (cmdline, argv[i]);
+		strcat (cmdline, " ");
+	}
```

Comment:
I'm not sure this would actually work, if `argv[i]` contains a space.

However, I think you can just do:

```suggestion
	char cmdline[MAX_PATH] = "";
	for (int i = 1; argv[i]; ++i)
	{
		strcat (cmdline, "\"");
		strcat (cmdline, argv[i]);
		strcat (cmdline, "\" ");
	}
```